torch/testing/_internal/common_methods_invocations.py

```diff
@@ -2535,6 +2535,34 @@ def sample_inputs_kron(op_info, device, dtype, requires_grad):
         sample_inputs.append(sample)
     return tuple(sample_inputs)
 
+def sample_inputs_scatter_ops(op_info, device, dtype, requires_grad):
+    def _make(shape, dtype=dtype, low=None, high=None):
+        return make_tensor(shape, device, dtype, low=low, high=high, requires_grad=requires_grad)
+
+    test_cases = (
```

Comment:
Hey @ysiraichi! Thanks for porting these operations to OpInfos.

This programmatic enumeration is neat, but it might be nicer to separate the sample input functions for these two operations and just hand-write most of the logic for now. In addition, what about testing the reduce parameter of scatter? 

So the code for scatter look like:

```
def sample_inputs_scatter(op_info, device, dtype, requires_grad, **kwargs):
  def _tensor(shape):
    return make_tensor(shape, device, dtype, requires_grad=requires_grad)
  def _gather(shape, index_dim, max_indices):
    return gather_variable(shape,  index_dim, max_indices, device=device)

  test_cases = (
    (_tensor((M, S)), 0, _gather((S, S), 1, M), _tensor((S, S))),
     ...
  )

  samples = []
  for test_case in test_cases:
    for reduce in (None, "add", "multiply"):
      sample = SampleInput(test_case[0], args=test_case[1:], kwargs={'reduce': reduce}
      samples.append(sample)

  return samples
```

scatter_add can reuse most of that function, except it doesn't have a reduce argument and (as you note in one of your comments) may not support all the inputs that scatter does.

What are you thoughts? Is an approach like this interesting? Would you suggest modifying it further?

Reply:
Sure! I guess making them more explicit might be better, indeed.
There is a thing with testing the reduce parameter of scatter, though. If I'm not mistaken, only its inplace variant `scatter_` supports it. Should I create another `OpInfo` specially for `scatter_`?

Reply:
Interesting! That's great attention to detail.

If only scatter_ supports the "reduce" parameter then I suggest not testing in this PR, actually, but filing an issue for this behavior ("scatter_ only supports 'reduce' kwarg on inplace variant") and the testing gap, and then adding a comment with a link to that issue above this sample input function. As for the sample input function itself, it would just remove the "reduce" enumeration, I think.

How does that sound? Then a future PR could look at extending the tests or extending scatter's kwarg support.

Reply:
That makes sense.
I will do that. Thanks!